### PR TITLE
[BACKLOG-37078] Focus Visibility and Focus Order for Browse Files (Folder Dialogs)

### PR DESCRIPTION
**Pull request by jsaraf**

### CODE REVIEW

**Comment by buildguy** [comment]
### :warning: Build finished in 5m 51s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl widgets
```
---
### :no_entry: Failed Tests

#### :cloud_with_lightning_and_rain: 1 test(s) failed:

<details><summary><b>org.pentaho.gwt.widgets.client.dialogs.PromptDialogBoxTest.testOnKeyDownPreview</b> (click to expand)</summary><p>

```

Wanted but not invoked:
promptDialogBox.onOk();
-> at org.pentaho.gwt.widgets.client.dialogs.PromptDialogBoxTest.testOnKeyDownPreview(PromptDialogBoxTest.java:36)

However, there were other interactions with this mock:
promptDialogBox.onKeyDownPreview('
', 0);
-> at org.pentaho.gwt.widgets.client.dialogs.PromptDialogBoxTest.testOnKeyDownPreview(PromptDialogBoxTest.java:35)


```

</p></details>

  
Tests run: **333**, Failures: **1**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/gwt-widgets/job/PR-838/1/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: widgets/src/main/java/org/pentaho/gwt/widgets/client/listbox/CustomListBox.java [hunk 1]
```diff
@@ -912,10 +912,16 @@ public void onFocus( Widget widget ) {
...
+    if ( visible == 1 ) { // drop-down mode
+      this.togglePopup();
+    }
     // fPanel.setFocus(true);
   }
 
   public void onLostFocus( Widget widget ) {
+    if( visible == 1 ){
+      popup.hide();
+    }
```
Comment (dcleao) [inline widgets/src/main/java/org/pentaho/gwt/widgets/client/listbox/CustomListBox.java:924]: While the ARIA combo-box pattern also admits the possibility of opening the dropdown when it receives focus, this is not, afaik, the default HTML behavior. In my system (MacOS), up, down or space, all open the popup and show the selected element.

File: widgets/src/main/java/org/pentaho/gwt/widgets/client/tabs/PentahoTab.java [hunk 2]
```diff
@@ -25,23 +25,26 @@
...
-public class PentahoTab extends SimplePanel {
+public class PentahoTab extends SimplePanel implements Focusable {
 
   public static final String SELECTED = "selected";
   private PentahoTabPanel tabPanel;
   private Widget content;
   protected Label label = new Label();
   private boolean solutionBrowserShowing;
 
+  private static final FocusImpl focusImpl = FocusImpl.getFocusImplForWidget();
```
Comment (dcleao) [inline widgets/src/main/java/org/pentaho/gwt/widgets/client/tabs/PentahoTab.java:47]: Please, add a new line after the field and before the constructor.

File: widgets/src/main/java/org/pentaho/gwt/widgets/client/tabs/PentahoTab.java [hunk 3]
```diff
@@ -185,4 +188,23 @@ public boolean isSolutionBrowserShowing() {
...
+  public void setFocus(boolean focus ) {
+    if (focus) {
+      focusImpl.focus(this.getElement());
+    }
+  }
+  @Override
+  public void setTabIndex(int index) {
+    focusImpl.setTabIndex(this.getElement(), index);
+  }
+
```
Comment (dcleao) [inline widgets/src/main/java/org/pentaho/gwt/widgets/client/tabs/PentahoTab.java:209]: Need a new line between methods. And no need for new line after the last one.
Also, spacing is off.

**Review by dcleao — CHANGES_REQUESTED** [review]

**Review by dcleao — CHANGES_REQUESTED** [review]
Additionally, there's a broken unit test.

File: widgets/src/main/java/org/pentaho/gwt/widgets/client/utils/ElementUtils.java [hunk 4]
```diff
@@ -245,7 +245,7 @@ public static Focusable findFirstKeyboardFocusableDescendant(Widget widget) {
       //TODO: the present implementation does not (yet) account for the disabled attribute of elements and that it
       // could be greatly optimized by not searching through parent elements hidden using display:none.
       Focusable focusable = ( Focusable ) widget;
-      if( focusable.getTabIndex() >= 0 && ElementUtils.isVisible( widget.getElement()) ) {
+      if( focusable.getTabIndex() >= 0 && ElementUtils.isVisible( widget.getElement()) && !widget.getElement().getPropertyBoolean("disabled")) {
```
Comment (dcleao) [inline widgets/src/main/java/org/pentaho/gwt/widgets/client/utils/ElementUtils.java:248]: Formatting is off.

**Comment by buildguy** [comment]
### :warning: Build finished in 6m 1s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl widgets
```
---
### :no_entry: Failed Tests

#### :cloud_with_lightning_and_rain: 1 test(s) failed:

<details><summary><b>org.pentaho.gwt.widgets.client.dialogs.PromptDialogBoxTest.testOnKeyDownPreview</b> (click to expand)</summary><p>

```

Wanted but not invoked:
promptDialogBox.onOk();
-> at org.pentaho.gwt.widgets.client.dialogs.PromptDialogBoxTest.testOnKeyDownPreview(PromptDialogBoxTest.java:36)

However, there were other interactions with this mock:
promptDialogBox.onKeyDownPreview('
', 0);
-> at org.pentaho.gwt.widgets.client.dialogs.PromptDialogBoxTest.testOnKeyDownPreview(PromptDialogBoxTest.java:35)


```

</p></details>

  
Tests run: **333**, Failures: **1**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/gwt-widgets/job/PR-838/2/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 6m 8s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl widgets
```
---
:ok_hand: All tests passed!
 
Tests run: **333**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/gwt-widgets/job/PR-838/3/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: widgets/src/main/java/org/pentaho/gwt/widgets/client/listbox/CustomListBox.java [hunk 5]
```diff
@@ -1027,8 +1030,11 @@ public void onKeyUp( Widget widget, char c, int i ) {
         break;
       case 27: // ESC
       case 13: // Enter
-        if ( popupShowing ) {
-          togglePopup();
+      case 32: // Space
+        if ( !popupShowing ) {
+          this.togglePopup();
+        } else {
+          this.popup.hide();
```
Comment (dcleao) [inline widgets/src/main/java/org/pentaho/gwt/widgets/client/listbox/CustomListBox.java:1037]: Just calling `this.togglePopup();` would do, for both when the popup is showing and not.
This method already checks for `popupShowing` and hides the popup when showing.

**Review by dcleao — CHANGES_REQUESTED** [review]

File: widgets/src/main/java/org/pentaho/gwt/widgets/client/listbox/CustomListBox.java (hunk 5 above)
Comment (dcleao) [reply widgets/src/main/java/org/pentaho/gwt/widgets/client/listbox/CustomListBox.java:1037]: Also, one other issue is, why would pressing ESC open the popup?
You should probably split the handling of these keys.

```java
  case 27: // ESC
     if ( popupShowing ) {
       togglePopup();
    }
    break;
  
  case 13: // Enter
  case 32: // Space
    this.togglePopup();
    break;
```

Comment (jsaraf) [reply widgets/src/main/java/org/pentaho/gwt/widgets/client/listbox/CustomListBox.java:1037]: > One other issue is, why would pressing ESC open the popup? You should probably split the handling of these keys.
> 
> ```java
>   case 27: // ESC
>     if ( popupShowing ) {
>       togglePopup();
>     }
>     break;
>   
>   case 13: // Enter
>   case 32: // Space
>     this.togglePopup();
>     break;
> ```

This line was already present in the code, anyway I've tested after removing the Esc case makes no difference in the behaviour (closing the dialog).

**Comment by jsaraf** [comment]
@dcleao Please review the changes once, thank you.

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 5m 37s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl widgets
```
---
:ok_hand: All tests passed!
 
Tests run: **333**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/gwt-widgets/job/PR-838/4/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: widgets/src/main/java/org/pentaho/gwt/widgets/client/listbox/CustomListBox.java (hunk 5 above)
Comment (dcleao) [reply widgets/src/main/java/org/pentaho/gwt/widgets/client/listbox/CustomListBox.java]: Before, ESC was only taking effect when hiding the popup. Your proposed code changed started opening the popup when hidden.

Comment (dcleao) [reply widgets/src/main/java/org/pentaho/gwt/widgets/client/listbox/CustomListBox.java]: And the newly proposed code completely removed the ESC key handling. AFAIK, the ESC key should close the dialog.

Above, I proposed:
```java
case 27: // ESC
    if ( popupShowing ) {
      this.togglePopup();
    }
    break;
```

But you did not include that.

Comment (jsaraf) [reply widgets/src/main/java/org/pentaho/gwt/widgets/client/listbox/CustomListBox.java]: There was an old bug which I was trying to troubleshoot for which I had to create an override method. Will be adding that code change back like you suggested.

File: widgets/src/main/java/org/pentaho/gwt/widgets/client/listbox/CustomListBox.java [hunk 6]
```diff
@@ -1025,10 +1028,14 @@ public void onKeyUp( Widget widget, char c, int i ) {
           }
         }
         break;
-      case 27: // ESC
+      case 27: // Esc
+        break;
```
Comment (dcleao) [inline widgets/src/main/java/org/pentaho/gwt/widgets/client/listbox/CustomListBox.java:1032]: By adding the new code, this _case_ is not needed then, here.

**Review by dcleao — CHANGES_REQUESTED** [review]

File: widgets/src/main/java/org/pentaho/gwt/widgets/client/listbox/CustomListBox.java (hunk 6 above)
Comment (jsaraf) [reply widgets/src/main/java/org/pentaho/gwt/widgets/client/listbox/CustomListBox.java:1032]: Okay, will be redacting the "Escape" case

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 5m 46s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl widgets
```
---
:ok_hand: All tests passed!
 
Tests run: **333**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/gwt-widgets/job/PR-838/5/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by jsaraf** [comment]
@dcleao Changes done according to suggested "final" code.

**Review by ddiroma — APPROVED** [review]

**Review by dcleao — APPROVED** [review]

**Pull request merged**